### PR TITLE
GLUE/SuperGLUE Tasks Improvements

### PR DESCRIPTION
**Pull request by slowwavesleep**
## Features

This PR deals mainly with the implementation/reproduction of various tasks from GLUE and SuperGLUE benchmarks, as well as introducing the more recent version of HuggingFace's transformers to the library.

- Bump `transformers` version from 2.8.0 to 4.6.0
- Compatibility fixes to various tasks related to the `transformers` version update
- Add RTE and MNLI configs
- Add a new task type and a corresponding pipeline: Multiple Choice
- Add helper functions and config for the Choice of Plausible Alternatives task (COPA)
- Revise affected NER and SQuAD pipelines and configs to accommodate them to the `transformers` update
- Add new pre-trained model download paths where applicable
- Update model scores in the docs
- Experimental multi-GPU support

### CODE REVIEW

**Comment by yurakuratov** [comment]
Could you, please, add PR description and provide metrics (in docs or here) for new and updated models?

**Comment by vaskonov** [comment]
> Could you, please, add PR description and provide metrics (in docs or here) for new and updated models?

The scores were updated in the docs

**Comment by danielkornev** [comment]
**Triage for v0.16**
# Sberbank 
- [x] - Add RTE and MNLI configs
- [x] - Add helper functions and config for the Choice of Plausible Alternatives task (COPA)
- [ ] - BoolQ | staged for v0.17

# Refactoring Library & Specific Existing Models
- [x] - Revise affected NER and SQuAD pipelines and configs to accommodate them to the transformers update
- [x] - Update model scores in the docs
- [x] - Add new pre-trained model download paths where applicable
- [x] - Bump transformers version from 2.8.0 to 4.6.0
- [x] - Compatibility fixes to various tasks related to the transformers version update

# New Features
- [x] - Add a new task type and a corresponding pipeline: Multiple Choice

# New Experimental Features
- [ ] - Experimental multi-GPU support | staged for v0.17

File: deeppavlov/configs/squad/squad_torch_bert.json [hunk 1]
```diff
@@ -158,16 +158,18 @@
...
       "ROOT_PATH": "~/.deeppavlov",
       "DOWNLOADS_PATH": "{ROOT_PATH}/downloads",
-      "MODELS_PATH": "{ROOT_PATH}/models/squad_torch_bert"
+      "MODELS_PATH": "{ROOT_PATH}/models",
+      "MODEL_PATH": "{MODELS_PATH}/squad_torch_bert/{TRANSFORMER}"
     },
     "download": [
       {
-        "url": "http://files.deeppavlov.ai/deeppavlov_data/squad_torch_bert_v0.tar.gz",
+        "url": "https://files.deeppavlov.ai/v1/squad/squad_torch_bert.tar.gz",
```
Comment (ayeffkay) [inline deeppavlov/configs/squad/squad_torch_bert.json:170]: Is this file with new pretrained bert using transformers 4.6.0? I've also trained squad torch bert for compatibility, here is url: http://files.deeppavlov.ai/deeppavlov_data/squad_torch_bert_v1.tar.gz
I'll change url in my config to yours and remove my archive, ok?

Comment (yurakuratov) [reply deeppavlov/configs/squad/squad_torch_bert.json:170]: Evaluation scores should be also considered to make this decision.

File: deeppavlov/models/embedders/transformers_embedder.py [hunk 2]
```diff
@@ -52,7 +52,9 @@ def load(self):
         self.model = transformers.BertModel.from_pretrained(self.load_path, config=self.config).eval().to(self.device)
```
Comment (ayeffkay) [inline deeppavlov/models/embedders/transformers_embedder.py:52]: `return_dict=False` should be added here, I had errors in `sentiment_twitter_bert_emb.json` and `ner_ontonotes_bert_emb.json` during `deeppavlov train ...` because code returns instance of ` BaseModelOutputWithPoolingAndCrossAttentions` instead of tuple
or I can rewrite my code as yours

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py [hunk 3]
```diff
@@ -31,6 +31,88 @@
...
+        tokenizer: instance of Bert FullTokenizer
+
+    """
+
+    def __init__(self,
+                 vocab_file: str,
+                 do_lower_case: bool = True,
+                 max_seq_length: int = 512,
+                 return_tokens: bool = False,
+                 **kwargs) -> None:
```
Comment (ayeffkay) [inline deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:58]: Please, implement adding token_type_ids inside `tokenize_mc_examples`, if tokenizer does not generate it. Distil* tokenizer does not return `token_type_ids`, and this causes problems with training and evaluation. 
You should add  argument add_token_type_ids: bool = False (as done in my code) and generate vector with 0s for `text` and 1s for `text_pair` after tokenization

File: deeppavlov/models/embedders/transformers_embedder.py (hunk 2 above)
Comment (slowwavesleep) [reply deeppavlov/models/embedders/transformers_embedder.py:52]: See my comments regarding this issue in your PR. My opinion is that it's preferable to use these dict-like `ModelOutput` structures because they're a bit more intuitive to work with (getting loss with `.loss` instead of `[0]` is much more concise and readable). Moreover, I have other changes in this PR that are dependent on these structures.

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py (hunk 3 above)
Comment (slowwavesleep) [reply deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:58]: The newer tokenizer does that on it's own depending on the pretrained model. For example, tokenizer from pretrained bert will return token_type_ids, while tokenizer from pretrained roberta won't. You don't have to do all of this.

Comment (slowwavesleep) [reply deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:58]: The problem might be that you're iterating over hard-coded list of features in `__call__` or `forward`.

File: deeppavlov/models/torch_bert/torch_transformers_squad.py [hunk 4]
```diff
@@ -124,10 +137,12 @@ def train_on_batch(self, features: List[InputFeatures], y_st: List[List[int]], y
...
-        outputs = self.model(input_ids=b_input_ids, attention_mask=b_input_masks,
-                             token_type_ids=b_input_type_ids,
-                             start_positions=b_y_st, end_positions=b_y_end)
-        loss = outputs[0]
+        loss = self.model(input_ids=b_input_ids,
+                          attention_mask=b_input_masks,
+                          token_type_ids=b_input_type_ids,
+                          start_positions=b_y_st,
+                          end_positions=b_y_end,
+                          return_dict=True).loss
```
Comment (ayeffkay) [inline deeppavlov/models/torch_bert/torch_transformers_squad.py:145]: Please, add check if model.forward has all passed arguments (as you did in `torch_transformers_classifier`). Distil* class hasn't `token_type_ids` argument, this code will raise error.

File: deeppavlov/configs/squad/squad_torch_bert.json (hunk 1 above)
Comment (slowwavesleep) [reply deeppavlov/configs/squad/squad_torch_bert.json:170]: Yes, it's a model that was retrained 4.6.0, because the weights trained on the old version couldn't be loaded properly. And it performs a bit worse, although it's not clear why. I'll have to check whether the weights that you're suggesting are going to load/work on my branch first.

File: deeppavlov/models/torch_bert/torch_transformers_squad.py (hunk 4 above)
Comment (slowwavesleep) [reply deeppavlov/models/torch_bert/torch_transformers_squad.py:145]: Do you checking the accepted_keys property? Like this?
```
tokenized = {key: value for (key, value) in _input.items() if key in self.accepted_keys}
```

Comment (ayeffkay) [reply deeppavlov/models/torch_bert/torch_transformers_squad.py:145]: I mean `if arg in model.forward.__code__.co_varnames`. Yes, it can be implemented using `accepted_keys` property.

File: deeppavlov/models/preprocessors/torch_transformers_preprocessor.py (hunk 3 above)
Comment (ayeffkay) [reply deeppavlov/models/preprocessors/torch_transformers_preprocessor.py:58]: No, it's problem with all Distil class. It doesn't work with `token_type_ids`

**Review by ayeffkay — APPROVED** [review]
